Build problem: same code build and works with 1.4.0 but it doesn't with 1.4.1

## Expected Behavior

It was working in 1.4.0, I would say that it should work the same in 1.4.1.

## Actual Behavior

It builds and works with 1.4.0, but it does not build anymore with 1.4.1.

```bash
include/hpx/traits/pack_traversal_rebind_container.hpp:70:9: error: base class has incomplete type
      : pack_traversal_rebind_container<NewType, OldType>
```

I played with it a bit, it is not related to the unwrapping, it happens even just with the dataflow, as soon as I `.get()` the future moved in, i.e. I can do whatever I want but not use the future passed in.

## Steps to Reproduce the Problem

See the code below, it has a switch flag `ENABLE_CUSTOM`.

With 1.4.0 both versions build and work.
With 1.4.1 the ENABLE_CUSTOM version with `custom::vector` does not build.

```cpp
#include <iostream>
#include <vector>

#include <hpx/hpx.hpp>
#include <hpx/hpx_init.hpp>

namespace custom {

template <class T>
struct vector : public std::vector<T> {
  using std::vector<T>::vector;

  vector(int size) : vector<T>(static_cast<std::size_t>(size)) {
    std::cout << "custom" << std::endl;
  }
};

}

int start(int argc, char **argv) {
  using T = int;

#ifdef ENABLE_CUSTOM
  using custom::vector;
#else
  using std::vector;
#endif

  vector<hpx::future<T>> values_futures;

  for (int i = 1; i <= 3; ++i)
    values_futures.emplace_back(hpx::make_ready_future<T>(i));

  auto sum = hpx::dataflow(
      hpx::util::unwrapping([](const auto&& values) {
        return std::accumulate(values.begin(), values.end(), 0);
      }), values_futures).get();

  std::cout << sum << std::endl;

  return hpx::finalize();
}

int main(int argc, char **argv) {
  auto ret_code = hpx::init(start, argc, argv);

  return ret_code;
}
```

## Specifications

  - HPX Version: 1.4.1
  - Platform (compiler, OS): AppleClang 10.0.1. on MacOS 10.14.6 "Mojave"

--- a/tests/regressions/lcos/CMakeLists.txt
+++ b/tests/regressions/lcos/CMakeLists.txt
@@ -16,6 +16,7 @@ set(tests
     channel_not_empty_2890
     channel_register_as_2722
     dataflow_791
+    dataflow_4436
     dataflow_action_2008
     dataflow_const_functor_773
     dataflow_future_swap2

--- a/tests/regressions/lcos/dataflow_4436.cpp
+++ b/tests/regressions/lcos/dataflow_4436.cpp
@@ -1,0 +1,64 @@
+//  Copyright (c) 2020 albestro
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <hpx/hpx.hpp>
+#include <hpx/hpx_init.hpp>
+#include <hpx/testing.hpp>
+#include <hpx/traits/pack_traversal_rebind_container.hpp>
+
+#include <cstddef>
+#include <numeric>
+#include <vector>
+
+namespace custom {
+
+    template <typename T>
+    struct vector : public std::vector<T>
+    {
+        vector() = default;
+    };
+
+}    // namespace custom
+
+namespace hpx { namespace traits {
+
+    template <typename NewType, typename OldType>
+    struct pack_traversal_rebind_container<NewType, custom::vector<OldType>>
+    {
+        static custom::vector<NewType> call(custom::vector<OldType> const&)
+        {
+            // Create a new version of the container for the new data type
+            return custom::vector<NewType>();
+        }
+    };
+}}    // namespace hpx::traits
+
+int start(int argc, char** argv)
+{
+    custom::vector<hpx::future<int>> values_futures;
+
+    for (int i = 1; i <= 3; ++i)
+    {
+        values_futures.emplace_back(hpx::make_ready_future(i));
+    }
+
+    auto f = hpx::dataflow(hpx::util::unwrapping(
+        [](const auto&& values) {
+            return std::accumulate(values.begin(), values.end(), 0);
+        }),
+        values_futures);
+
+    HPX_TEST_EQ(f.get(), 6);
+
+    return hpx::finalize();
+}
+
+int main(int argc, char** argv)
+{
+    HPX_TEST_EQ(hpx::init(start, argc, argv), 0);
+
+    return hpx::util::report_errors();
+}

